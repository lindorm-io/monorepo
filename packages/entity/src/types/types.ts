export type IndexDirection =
  | "asc"
  | "desc"
  | "2d"
  | "2dsphere"
  | "geoHaystack"
  | "hashed"
  | "text";
