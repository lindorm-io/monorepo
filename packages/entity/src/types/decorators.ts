/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { ZodType } from "zod";
import {
  MetaColumnFallback,
  MetaColumnType,
  MetaGeneratedStrategy,
  MetaIndexOptions,
} from "./metadata";
import { IndexDirection } from "./types";

export type ColumnDecoratorOptions = {
  enum?: any;
  fallback?: MetaColumnFallback;
  max?: number;
  min?: number;
  nullable?: boolean;
  optional?: boolean;
  readonly?: boolean;
  schema?: ZodType;
  type?: MetaColumnType;
};

export type EntityDecoratorOptions = {
  cache?: string;
  database?: string;
  name?: string;
  namespace?: string;
};

export type GeneratedDecoratorOptions = {
  strategy?: MetaGeneratedStrategy;
  length?: number;
  max?: number;
  min?: number;
};

export type HookDecoratorCallback = (entity: any) => void;

export type IndexDecoratorOptions = {
  direction?: IndexDirection;
  name?: string;
  options?: MetaIndexOptions;
  unique?: boolean;
};

export type UniqueDecoratorOptions = {
  name?: string;
};
