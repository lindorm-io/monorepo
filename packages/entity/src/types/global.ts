/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { GlobalEntityMetadata } from "../classes/private";
import {
  MetaColumn,
  MetaEntity,
  MetaExtra,
  MetaGenerated,
  MetaHook,
  MetaIndex,
  MetaPrimaryKey,
  MetaPrimarySource,
  MetaSchema,
} from "./metadata";

export type GlobalMetadata = {
  columns: Array<MetaColumn>;
  entities: Array<MetaEntity>;
  extras: Array<MetaExtra>;
  generated: Array<MetaGenerated>;
  hooks: Array<MetaHook>;
  indexes: Array<MetaIndex>;
  primaryKeys: Array<MetaPrimaryKey>;
  primarySources: Array<MetaPrimarySource>;
  schemas: Array<MetaSchema>;
};

export type GlobalThisEntity = typeof globalThis & {
  __lindorm_entity__: GlobalEntityMetadata;
};
