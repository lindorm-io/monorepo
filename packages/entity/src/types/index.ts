export * from "./decorators";
export * from "./entity-utility";
export * from "./global";
export * from "./metadata";
export * from "./scanner";
export * from "./types";
