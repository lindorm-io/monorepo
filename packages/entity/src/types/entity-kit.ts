import { ILogger } from "@lindorm/logger";
import { Constructor } from "@lindorm/types";
import { IEntity } from "../interfaces";

export type GetIncrementFn = (key: string) => Promise<number>;

export type EntityKitOptions<E extends IEntity> = {
  Entity: Constructor<E>;
  logger?: ILogger;
  source: string;
  getNextIncrement?: GetIncrementFn;
};

export type NamespaceOptions = {
  namespace?: string | null;
};

export type SaveStrategy = "insert" | "update" | "unknown";

export type UpdateStrategy = "update" | "version";
