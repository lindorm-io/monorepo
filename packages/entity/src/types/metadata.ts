/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { Constructor, Dict } from "@lindorm/types";
import { ZodObject, ZodType } from "zod";
import { IEntity } from "../interfaces";
import { IndexDirection } from "./types";

export type MetaColumnType =
  | "array"
  | "bigint"
  | "boolean"
  | "date"
  | "email"
  | "enum"
  | "float"
  | "integer"
  | "object"
  | "string"
  | "uuid";

export type MetaColumnDecorator =
  // default
  | "Column"
  // special (unique)
  | "PrimaryKeyColumn"
  | "ScopeColumn"
  | "VersionColumn"
  // date
  | "CreateDateColumn"
  | "UpdateDateColumn"
  | "ExpiryDateColumn"
  | "DeleteDateColumn"
  // versioning
  | "VersionKeyColumn"
  | "VersionStartDateColumn"
  | "VersionEndDateColumn";

export type ColumnFallbackPrimitive = bigint | boolean | Date | number | string | null;
export type ColumnFallbackComplex =
  | Array<ColumnFallbackPrimitive>
  | Dict<ColumnFallbackPrimitive>
  | ColumnFallbackPrimitive;
export type MetaColumnFallback = (() => ColumnFallbackComplex) | ColumnFallbackComplex;

export type MetaColumn<T extends MetaColumnDecorator = MetaColumnDecorator> = {
  target: Function;
  key: string;
  decorator: T;
  enum: any | null; // enum corresponds to the type "enum" property
  fallback: MetaColumnFallback | null; // when value is not set, use this fallback value or function
  max: number | null; // max size (number) or length (array, string)
  min: number | null; // min size (number) or length (array, string)
  nullable: boolean; // can the column value be null
  optional: boolean; // can the column value be undefined
  readonly: boolean; // designates as readonly for operations where updates do not occur on the entity such as updateMany(criteria, partial)
  schema: ZodType | null; // overrides the type with a zod schema
  type: MetaColumnType | null; // used to determine zod schema for validation
};

export type MetaEntity = {
  target: Function;
  decorator: string;
  cache: string | null;
  database: string | null;
  name: string | null;
  namespace: string | null;
};

export type MetaExtra<T extends Dict = Dict> = {
  target: Function;
  type: string;
} & T;

export type MetaGeneratedStrategy =
  // basic
  | "date"
  | "float"
  | "integer"
  | "string"
  | "uuid"
  // special
  | "increment";

export type MetaGenerated = {
  target: Function;
  key: string;
  length: number | null;
  max: number | null;
  min: number | null;
  strategy: MetaGeneratedStrategy;
};

export type MetaHookDecorator =
  // before
  | "OnCreate"
  | "OnValidate"
  // after
  | "OnInsert"
  | "OnUpdate"
  | "OnDestroy";

export type MetaHook = {
  target: Function;
  decorator: MetaHookDecorator;
  callback: (entity: any) => void;
};

export type MetaIndexItem = {
  key: string;
  direction: IndexDirection;
};

export type MetaIndexOptions = {
  background?: boolean;
  bits?: number;
  bucketSize?: number;
  expireAfterSeconds?: number;
  hidden?: boolean;
  max?: number;
  min?: number;
  sparse?: boolean;
  version?: number;
};

export type MetaIndex = {
  target: Function;
  keys: Array<MetaIndexItem>;
  name: string | null;
  options: MetaIndexOptions;
  unique: boolean;
};

export type MetaPrimaryKey = {
  target: Function;
  key: string;
};

export type RelationChange = "cascade" | "default" | "restrict";
export type RelationLoading = "eager" | "lazy";
export type RelationOrphan = "delete" | "ignore";
export type RelationStrategy = "join" | "query";

export type MetaRelationOptions = {
  joinKey: boolean | string | null;
  joinTable: boolean | string | null;
  loading: RelationLoading | null;
  nullable: boolean;
  onDelete: RelationChange | null;
  onOrphan: RelationOrphan | null;
  onUpdate: RelationChange | null;
  strategy: RelationStrategy | null;
};

export type MetaRelation = {
  target: Function;
  key: string;
  options: MetaRelationOptions;
  foreignConstructor: () => Constructor<IEntity>;
  foreignKey: string | null;
  type: "ManyToMany" | "ManyToOne" | "OneToMany" | "OneToOne";
};

export type MetaSource =
  | "elastic"
  | "external"
  | "mnemos"
  | "mongo"
  | "postgres"
  | "redis";

export type MetaPrimarySource<T extends MetaSource = MetaSource> = {
  target: Function;
  source: T;
};

export type MetaSchema = {
  target: Function;
  schema: ZodObject<IEntity>;
};

export type EntityMetadata<
  TExtra extends Dict = Dict,
  TDecorator extends MetaColumnDecorator = MetaColumnDecorator,
  TSource extends MetaSource = MetaSource,
> = {
  columns: Array<Omit<MetaColumn<TDecorator>, "target">>;
  entity: Omit<MetaEntity, "target">;
  extras: Array<Omit<MetaExtra<TExtra>, "target">>;
  generated: Array<Omit<MetaGenerated, "target">>;
  hooks: Array<Omit<MetaHook, "target">>;
  indexes: Array<Omit<MetaIndex, "target">>;
  primaryKeys: Array<string>;
  primarySource: TSource | null;
  relations: Array<Omit<MetaRelation, "target">>;
  schemas: Array<ZodObject<IEntity>>;
};
