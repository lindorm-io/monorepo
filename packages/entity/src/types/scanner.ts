import { Constructor, Dict } from "@lindorm/types";

export type EntityScannerInput<T extends Dict = Dict> = Array<Constructor<T> | string>;
