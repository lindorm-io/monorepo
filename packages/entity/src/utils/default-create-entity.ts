import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

export const defaultCreateEntity = <
  E extends IEntity,
  O extends DeepPartial<E> = DeepPartial<E>,
>(
  Entity: Constructor<E>,
  options: O | E = {} as O,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const entity = new Entity();

  for (const column of metadata.columns) {
    (entity as any)[column.key] = parseColumn(column, entity, options);

    if ((entity as any)[column.key]) continue;

    (entity as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return entity;
};
