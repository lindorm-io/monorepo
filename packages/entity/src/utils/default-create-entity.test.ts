import { Dict } from "@lindorm/types";
import MockDate from "mockdate";
import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  Entity,
  ExpiryDateColumn,
  Generated,
  OnCreate,
  PrimaryKeyColumn,
  UpdateDateColumn,
  VersionColumn,
  VersionEndDateColumn,
  VersionKeyColumn,
  VersionStartDateColumn,
} from "../decorators";
import { defaultCreateEntity } from "./default-create-entity";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

jest.mock("crypto", () => ({
  ...jest.requireActual("crypto"),
  randomUUID: jest
    .fn()
    .mockImplementation(() => "c8ff3952-8ba4-51a3-a4d5-2f0726c49524")
    .mockImplementationOnce(() => "e8edbe03-f52f-56c6-8d98-82e6961329fe")
    .mockImplementationOnce(() => "0c8c31df-f1cc-5381-97fc-31b3714327de")
    .mockImplementationOnce(() => "2dc92d72-9ce3-50c7-b982-59022ec393f1")
    .mockImplementationOnce(() => "88943033-fef1-5ad8-8cf8-4470be80d197")
    .mockImplementationOnce(() => "1d37a07d-4fd6-5913-85eb-2c7d721cf2a9")
    .mockImplementationOnce(() => "5f2976d2-afda-5e28-9c06-3c04865ea8df"),
}));

describe("defaultCreateEntity", () => {
  test("should create with empty values", () => {
    enum TestEnum {
      One = "1",
      Two = "2",
    }

    @Entity()
    @OnCreate((entity) => {
      entity.OnCreate = "OnCreate";
    })
    class TestCreateEntityEmpty {
      @PrimaryKeyColumn()
      primaryKey!: string;

      @CreateDateColumn()
      createdDate!: Date;

      @DeleteDateColumn()
      deletedDate!: Date | null;

      @ExpiryDateColumn()
      expiryDate!: Date | null;

      @Column()
      @Generated("increment")
      sequence!: number;

      @UpdateDateColumn()
      updatedDate!: Date;

      @VersionColumn()
      version!: number;

      @VersionEndDateColumn()
      versionEndDate!: Date | null;

      @VersionKeyColumn()
      versionKey!: string;

      @VersionStartDateColumn()
      versionStartDate!: Date;

      @Column("array")
      array!: any[];

      @Column("bigint")
      bigint!: bigint;

      @Column("boolean")
      boolean!: boolean;

      @Column("date")
      date!: Date;

      @Column("enum", { enum: TestEnum })
      enum!: TestEnum;

      @Column("integer")
      number!: number;

      @Column("object")
      object!: Dict;

      @Column("string")
      string!: string;

      @Column("uuid")
      uuid!: string;
    }

    expect(defaultCreateEntity(TestCreateEntityEmpty)).toMatchSnapshot();
  });

  test("should create with fallback values", () => {
    enum TestEnum {
      One = "1",
      Two = "2",
    }

    @Entity()
    class TestCreateEntityFallback {
      @PrimaryKeyColumn()
      id!: string;

      @Column("array", { fallback: [] })
      array!: any[];

      @Column("bigint", { fallback: BigInt(0) })
      bigint!: bigint;

      @Column("boolean", { fallback: false })
      boolean!: boolean;

      @Column("date", { fallback: () => new Date() })
      date!: Date;

      @Column("enum", { enum: TestEnum, fallback: TestEnum.One })
      enum!: TestEnum;

      @Column("integer", { fallback: 0 })
      number!: number;

      @Column("object", { fallback: {} })
      object!: Dict;

      @Column("string", { fallback: "string" })
      string!: string;

      @Column("uuid", { fallback: "00000000-0000-0000-0000-000000000000" })
      uuid!: string;
    }

    expect(defaultCreateEntity(TestCreateEntityFallback)).toMatchSnapshot();
  });

  test("should create with options values", () => {
    enum TestEnum {
      One = "1",
      Two = "2",
    }

    @Entity()
    class TestCreateEntityOptions {
      @PrimaryKeyColumn()
      id!: string;

      @Column("array", { fallback: [] })
      array!: any[];

      @Column("bigint", { fallback: BigInt(0) })
      bigint!: bigint;

      @Column("boolean", { fallback: false })
      boolean!: boolean;

      @Column("date", { fallback: new Date() })
      date!: Date;

      @Column("enum", { enum: TestEnum, fallback: TestEnum.One })
      enum!: TestEnum;

      @Column("integer", { fallback: 0 })
      number!: number;

      @Column("object", { fallback: {} })
      object!: Dict;

      @Column("string", { fallback: "string" })
      string!: string;

      @Column("uuid", { fallback: "00000000-0000-0000-0000-000000000000" })
      uuid!: string;
    }

    expect(
      defaultCreateEntity(TestCreateEntityOptions, {
        array: [1, 2, 3],
        bigint: BigInt(1),
        boolean: true,
        date: new Date("2025-01-01T00:00:00.000Z"),
        enum: TestEnum.Two,
        number: 1,
        object: { key: "value" },
        string: "value",
        uuid: "00000000-0000-0000-0000-000000000001",
      }),
    ).toMatchSnapshot();
  });
});
