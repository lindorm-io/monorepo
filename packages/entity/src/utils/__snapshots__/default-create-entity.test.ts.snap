// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultCreateEntity should create with empty values 1`] = `
TestCreateEntityEmpty {
  "OnCreate": "OnCreate",
  "array": null,
  "bigint": null,
  "boolean": null,
  "createdDate": 2024-01-01T08:00:00.000Z,
  "date": null,
  "deletedDate": null,
  "enum": null,
  "expiryDate": null,
  "number": null,
  "object": null,
  "primaryKey": "e8edbe03-f52f-56c6-8d98-82e6961329fe",
  "sequence": null,
  "string": null,
  "updatedDate": 2024-01-01T08:00:00.000Z,
  "uuid": null,
  "version": 0,
  "versionEndDate": null,
  "versionKey": "0c8c31df-f1cc-5381-97fc-31b3714327de",
  "versionStartDate": 2024-01-01T08:00:00.000Z,
}
`;

exports[`defaultCreateEntity should create with fallback values 1`] = `
TestCreateEntityFallback {
  "array": [],
  "bigint": 0n,
  "boolean": false,
  "date": 2024-01-01T08:00:00.000Z,
  "enum": "1",
  "id": "2dc92d72-9ce3-50c7-b982-59022ec393f1",
  "number": 0,
  "object": {},
  "string": "string",
  "uuid": "00000000-0000-0000-0000-000000000000",
}
`;

exports[`defaultCreateEntity should create with options values 1`] = `
TestCreateEntityOptions {
  "array": [
    1,
    2,
    3,
  ],
  "bigint": 1n,
  "boolean": true,
  "date": 2025-01-01T00:00:00.000Z,
  "enum": "2",
  "id": "88943033-fef1-5ad8-8cf8-4470be80d197",
  "number": 1,
  "object": {
    "key": "value",
  },
  "string": "value",
  "uuid": "00000000-0000-0000-0000-000000000001",
}
`;
